File: CS/MvpvmNavigation/Services/ModuleLocator.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using DevExpress.Mvvm;
using MvpvmNavigation.Model;

namespace MvpvmNavigation.Services {
    public interface IModuleLocator {
        bool IsModuleLoaded(ModuleType moduleType);
        object GetModule(ModuleType moduleType);
        void ReleaseModule(object module);
    }
    //
    class ModuleLocator : IModuleLocator {
        IServiceContainer serviceContainer;
        IDictionary<ModuleType, WeakReference> modulesCache;
        public ModuleLocator(DevExpress.Mvvm.IServiceContainer serviceContainer) {
            this.serviceContainer = serviceContainer;
            this.modulesCache = new Dictionary<ModuleType, WeakReference>();
        }
        public bool IsModuleLoaded(ModuleType moduleType) {
            return modulesCache.ContainsKey(moduleType);
        }
        public object GetModule(ModuleType moduleType) {
            if(moduleType == ModuleType.None)
                return null;
            WeakReference moduleReference;
            if(!modulesCache.TryGetValue(moduleType, out moduleReference) || moduleReference.Target == null) {
                var resolver = serviceContainer.GetService<IModuleTypesResolver>();
                string moduleTypeName = resolver.GetTypeName(moduleType);
                var activator = serviceContainer.GetService<IModuleActivator>();
                var module = activator.CreateModule(moduleTypeName);
                if(moduleReference == null) {
                    moduleReference = new WeakReference(module);
                    modulesCache.Add(moduleType, moduleReference);
                }
                else moduleReference.Target = module;
            }
            return moduleReference.Target;
        }
        public void ReleaseModule(object module) {
            ModuleType key = ModuleType.None;
            foreach(var item in modulesCache) {
                if(!object.Equals(item.Value.Target, module)) continue;
                key = item.Key;
                break;
            }
            modulesCache.Remove(key);
        }
    }
}

File: VB/MvpvmNavigation/Services/ModuleLocator.vb
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Text
Imports System.Threading.Tasks
Imports DevExpress.Mvvm
Imports MvpvmNavigation.Model

Namespace MvpvmNavigation.Services
    Public Interface IModuleLocator
        Function IsModuleLoaded(ByVal moduleType As ModuleType) As Boolean
        Function GetModule(ByVal moduleType As ModuleType) As Object
        Sub ReleaseModule(ByVal [module] As Object)
    End Interface
    '
    Friend Class ModuleLocator
        Implements IModuleLocator

        Private serviceContainer As IServiceContainer
        Private modulesCache As IDictionary(Of ModuleType, WeakReference)
        Public Sub New(ByVal serviceContainer As DevExpress.Mvvm.IServiceContainer)
            Me.serviceContainer = serviceContainer
            Me.modulesCache = New Dictionary(Of ModuleType, WeakReference)()
        End Sub
        Public Function IsModuleLoaded(ByVal moduleType As ModuleType) As Boolean Implements IModuleLocator.IsModuleLoaded
            Return modulesCache.ContainsKey(moduleType)
        End Function
        Public Function GetModule(ByVal moduleType As ModuleType) As Object Implements IModuleLocator.GetModule
            If moduleType = MvpvmNavigation.Model.ModuleType.None Then
                Return Nothing
            End If
            Dim moduleReference As WeakReference = Nothing
            If (Not modulesCache.TryGetValue(moduleType, moduleReference)) OrElse moduleReference.Target Is Nothing Then
                Dim resolver = serviceContainer.GetService(Of IModuleTypesResolver)()
                Dim moduleTypeName As String = resolver.GetTypeName(moduleType)
                Dim activator = serviceContainer.GetService(Of IModuleActivator)()
                Dim [module] = activator.CreateModule(moduleTypeName)
                If moduleReference Is Nothing Then
                    moduleReference = New WeakReference([module])
                    modulesCache.Add(moduleType, moduleReference)
                Else
                    moduleReference.Target = [module]
                End If
            End If
            Return moduleReference.Target
        End Function
        Public Sub ReleaseModule(ByVal [module] As Object) Implements IModuleLocator.ReleaseModule
            Dim key As ModuleType = ModuleType.None
            For Each item In modulesCache
                If Not Object.Equals(item.Value.Target, [module]) Then
                    Continue For
                End If
                key = item.Key
                Exit For
            Next item
            modulesCache.Remove(key)
        End Sub
    End Class
End Namespace
